### ISSUE TITLE
FTP backend: Put, renaming. Still, the old filename is present at the backend

### ISSUE THREAD

**Issue by huste** [duplicati/duplicati#4126]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info
- **Duplicati version**: 2.0.5.103
- **Operating system**: Win10 Pro
- **Backend**: FTP Alternative

## Description
I use Duplicato on my father's slow internet line. As a backend, the only available thing is an FTP server via OpenVPN. The upload speed is limited to about 10KB/s, I am not using the duplicati-throttle-setting any more.

I noticed in the log, that Duplicati uploads a .zip file (19MB) and then renames them. Still, on the backend, I can see both files existing. I am not sure, what happens here. Can anyone enlighten me?

## Debug log
Here's an excerpt of the log file. You can see, there's a "Starting", "Retrying", "Rename" for several files (not all). But in this case, they all exist on the backend. 

2020-03-04 20:29:17 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-b9fff0244be064fcabe82e36f05860b1a.dblock.zip (19,91 MB)
2020-03-04 20:43:08 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Retrying: duplicati-b9fff0244be064fcabe82e36f05860b1a.dblock.zip (19,91 MB)
2020-03-04 20:43:18 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-b9fff0244be064fcabe82e36f05860b1a.dblock.zip (19,91 MB)
2020-03-04 20:43:18 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-bd0ac23cfdaf8410887aca719a952942a.dblock.zip (19,91 MB)
2020-03-04 20:43:18 +01 - [Information-Duplicati.Library.Main.Operation.Backup.BackendUploader-RenameRemoteTargetFile]: Renaming "duplicati-b9fff0244be064fcabe82e36f05860b1a.dblock.zip" to "duplicati-bd0ac23cfdaf8410887aca719a952942a.dblock.zip"
2020-03-04 20:43:18 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-bd0ac23cfdaf8410887aca719a952942a.dblock.zip (19,91 MB)
2020-03-04 20:57:06 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Completed: duplicati-bd0ac23cfdaf8410887aca719a952942a.dblock.zip (19,91 MB)
2020-03-04 20:57:06 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-i5cb4acc34afe42c88a2c11c3e7a0a875.dindex.zip (91,25 KB)
2020-03-04 20:57:16 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Completed: duplicati-i5cb4acc34afe42c88a2c11c3e7a0a875.dindex.zip (91,25 KB)
2020-03-04 20:57:16 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-b3750abe39aab4249881d12a0be5bd48c.dblock.zip (19,92 MB)
2020-03-04 21:11:47 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Retrying: duplicati-b3750abe39aab4249881d12a0be5bd48c.dblock.zip (19,92 MB)
2020-03-04 21:11:57 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-b3750abe39aab4249881d12a0be5bd48c.dblock.zip (19,92 MB)
2020-03-04 21:11:57 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-bd739df472ff44108ad60dda9392e1f71.dblock.zip (19,92 MB)
2020-03-04 21:11:57 +01 - [Information-Duplicati.Library.Main.Operation.Backup.BackendUploader-RenameRemoteTargetFile]: Renaming "duplicati-b3750abe39aab4249881d12a0be5bd48c.dblock.zip" to "duplicati-bd739df472ff44108ad60dda9392e1f71.dblock.zip"
2020-03-04 21:11:57 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-bd739df472ff44108ad60dda9392e1f71.dblock.zip (19,92 MB)
2020-03-04 21:26:51 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Retrying: duplicati-bd739df472ff44108ad60dda9392e1f71.dblock.zip (19,92 MB)
2020-03-04 21:27:01 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-bd739df472ff44108ad60dda9392e1f71.dblock.zip (19,92 MB)
2020-03-04 21:27:01 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-b8f4f959e48674340a4300239ae0d699f.dblock.zip (19,92 MB)
2020-03-04 21:27:01 +01 - [Information-Duplicati.Library.Main.Operation.Backup.BackendUploader-RenameRemoteTargetFile]: Renaming "duplicati-bd739df472ff44108ad60dda9392e1f71.dblock.zip" to "duplicati-b8f4f959e48674340a4300239ae0d699f.dblock.zip"
2020-03-04 21:27:01 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-b8f4f959e48674340a4300239ae0d699f.dblock.zip (19,92 MB)
2020-03-04 21:41:23 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Retrying: duplicati-b8f4f959e48674340a4300239ae0d699f.dblock.zip (19,92 MB)
2020-03-04 21:41:33 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-b8f4f959e48674340a4300239ae0d699f.dblock.zip (19,92 MB)
2020-03-04 21:41:33 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-b1bc44b09095d46c5bf143df0c979087b.dblock.zip (19,92 MB)
2020-03-04 21:41:33 +01 - [Information-Duplicati.Library.Main.Operation.Backup.BackendUploader-RenameRemoteTargetFile]: Renaming "duplicati-b8f4f959e48674340a4300239ae0d699f.dblock.zip" to "duplicati-b1bc44b09095d46c5bf143df0c979087b.dblock.zip"
2020-03-04 21:41:33 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-b1bc44b09095d46c5bf143df0c979087b.dblock.zip (19,92 MB)
2020-03-04 21:55:47 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Completed: duplicati-b1bc44b09095d46c5bf143df0c979087b.dblock.zip (19,92 MB)
2020-03-04 21:55:48 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-idf2c12f7083348e29fcc484474c60993.dindex.zip (221,35 KB)
2020-03-04 21:56:00 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Completed: duplicati-idf2c12f7083348e29fcc484474c60993.dindex.zip (221,35 KB)
2020-03-04 21:56:00 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Started: duplicati-b53cbebba8bff44b5ae219620b9c2d5ad.dblock.zip (19,93 MB)
2020-03-04 22:11:12 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Retrying: duplicati-b53cbebba8bff44b5ae219620b9c2d5ad.dblock.zip (19,93 MB)
2020-03-04 22:11:22 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-b53cbebba8bff44b5ae219620b9c2d5ad.dblock.zip (19,93 MB)
2020-03-04 22:11:22 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: Put - Rename: duplicati-bcbd257a380424975a02e9fce3e4d44d9.dblock.zip (19,93 MB)
2020-03-04 22:11:22 +01 - [Information-Duplicati.Library.Main.Operation.Backup.BackendUploader-RenameRemoteTargetFile]: Renaming "duplicati-b53cbebba8bff44b5ae219620b9c2d5ad.dblock.zip" to "duplicati-bcbd257a380424975a02e9fce3e4d44d9.dblock.zip"

**Comment by huste** [comment]
Addendum: I noticed, that not only the old and new files are present on the backend. Duplicati *uploads* the renamed file to the backend:
1) file of name A is uploaded
2) Duplicati logs, that A is renamed to B
3) file of name B is uploaded

**Comment by ts678** [comment]
> Duplicati uploads the renamed file to the backend

That's the right interpretation. Not every backend can rename after upload, but Duplicati does a retry of a failed upload under a different name for same content to work around a bug some destinations have. 

The failed (maybe partial) upload should be deleted at some point. I forget if it's at end, or at next start.
If you're interested, you can look in DB Remotevolume table to see files' State. It's probably `Uploading`.

[--number-of-retries](https://duplicati.readthedocs.io/en/latest/06-advanced-options/#number-of-retries) can be raised if necessary, but preferably you would get successful uploads instead of failures. Watching the live log in About --> Show log --> Live --> Retry might give more information. You can click on some of the messages to expand their error. How reliable is this slow VPN connection?

**Comment by huste** [comment]
Bummer. That's bad news. I can't tell for sure how reliable it is, but let me rephrase it:
* The connection is a very slow one (<15KB/s upload)
* The VPN does not log any reconnects during the whole operation.
* The backend FTP Server on my Synology NAS at home did not have logging enabled, I enabled it and will make a small backup (of about 400MB) and report back with a look at the FTP log, the OpenVPN log and the Duplicati log.

**Comment by ts678** [comment]
I suppose you could look for any patterns, for example does chance of failure increase with file size? 
You could try [FTP](https://duplicati.readthedocs.io/en/latest/05-storage-providers/#ftp) with --ftp-passive then with --ftp-regular. Check Duplicati logs for error messages.
You could see if setting --disable-upload-verify for FTP helps. Is remote file about the expected size? When multiple retries leave multiple remote files, are files all the same size or do their file sizes vary?

You could see if similarly sized files have any trouble with uploading using some other FTP programs.
You could download a file that got retried, and try to unzip if unencrypted, or decrypt with [AES Crypt](https://www.aescrypt.com/).

**Comment by huste** [comment]
ad file sizes: The chunks are set to 20MB, which each file has. That's due to the limited bandwitdh and to prevent large files being aborted.

ad FTP: I can't remember why I am using "FTP alternative" as backend , but "FTP" did not work at all and SMB/CIFS was way too slow. I think, I'll create a sample backup job with about only 100MB of data and try various variants with that.

ad disabling verification: disable-upload-verify is already set. Duplicati still fails.

yes, the remote files are *all* around 20MB. The current FTP log from the server lists 22 having been put. However, on the backend, there are 26 files. This is after *one* run of the backup job. The state of those files is "Deleting".

The FTP protocol does not list any aborts, the VPN is up an operative since a few days without interruptions. I am currently running a second try with the same settings as yesterday and I am already seeing the superfluous files being removed from the backend.

Attached is a ZIP file of the FTP log and Duplicati's error message that was sent to me. The log is unfortunately gone, I forgot that the c:\temp directory is wiped every day on my father's machine...


Question: Is the issue with not being able to rename after upload a mere question of the backend or a question of whether Duplicati does properly support it? If I'm not terribly mistaken, FTP does support the rename command..

I'll post the current logs after this job has finished.

**Comment by huste** [comment]
Addendum: I have created a simpler job with the FTP backend, using chunks of 5MB for each upload file. There are no retries and no renaming issues at the moment. I have changed the original backup job to use FTP (pasive mode) rather than FTP alternative . And let's see. Results will have to wait until tomorrow :)

**Comment by ts678** [comment]
Is there any network address translation (NAT) in the picture? FTP passes actual address and port information around, and NAT changes all of that (commonly in a home router going to Internet). Workaround is for the router to rewrite FTP's control channel messages, but this might time out.

I don't know how IP addressing between local and remote works under OpenVPN in your setup.

Though I'm not familiar with FluentFTP internals, this looks like possibly the upload finished, but attempting to read back the status through the control connection was getting a timeout error:

```
---> (Interne Ausnahme #1) System.AggregateException: Mindestens ein Fehler ist aufgetreten. ---> System.TimeoutException: Timed out trying to read data from the socket stream!
   bei FluentFTP.FtpSocketStream.<ReadAsync>d__68.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei FluentFTP.FtpSocketStream.<ReadLineAsync>d__71.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei FluentFTP.FtpClient.<GetReplyAsync>d__263.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei FluentFTP.FtpClient.<UploadFileInternalAsync>d__225.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei Duplicati.Library.Main.Operation.Backup.BackendUploader.<<Run>b__13_0>d.MoveNext()
```

Here's an example of what an FTP upload should look like. STOR is upload. The rest are responses.

```
STOR 2.0.5.1_ftp_test_1/duplicati-20200308T200316Z.dlist.zip
150 Opening data channel for file upload to server of "/2.0.5.1_ftp_test_1/duplicati-20200308T200316Z.dlist.zip"
226 Successfully transferred "/2.0.5.1_ftp_test_1/duplicati-20200308T200316Z.dlist.zip"
```

In terms of more experiments, leaving it alone as long as it keeps working is also a reasonable path.

**Comment by huste** [comment]
No, there's no NAT involved. The VPN connection terminates at the Synology which als serves as nbackup target, its just plain routing, as far as I get it.
I am continuously adding more directories to the backup job. So far, each day did work perfectly with the FTP backend (not the FTP alternative), No renamings, no stale connections, no other aborts.

I somehow refrain from using the production backup job for any further tests, so I'll stick to the FTP backend. But if I can help out with a smaller backup job if some testing needs to be done for the alternative FTP backend, I'll gladly perform them. Responses take very long, though :)

**Comment by ts678** [comment]
> Question: Is the issue with not being able to rename after upload a mere question of the backend or a question of whether Duplicati does properly support it? If I'm not terribly mistaken, FTP does support the rename command..

Duplicati doesn't use it, because it's written for a [minimal common set of capabilities](https://github.com/duplicati/duplicati/wiki/Developer-Documentation#overview) that all backends have. Rename isn't one.
After a failed upload, it doesn't even make sense to instantly rename. There might not even be a file there. Previous note said:

> Not every backend can rename after upload, but Duplicati does a retry of a failed upload under a different name

> The state of those files is "Deleting".

Files in that state are flagged in the database for deletion later on, probably by the code below which cleans up unwanted files:

https://github.com/duplicati/duplicati/blob/e7c191389cdb3fdb9500d83ebd06e0ae99634fd1/Duplicati/Library/Main/Operation/FilelistProcessor.cs#L51-L58

https://github.com/duplicati/duplicati/blob/e7c191389cdb3fdb9500d83ebd06e0ae99634fd1/Duplicati/Library/Main/Enums.cs#L80-L83

> So far, each day did work perfectly with the FTP backend (not the FTP alternative), No renamings, no stale connections, no other aborts.

So that last post was good operating news, and I've answered questions a bit further now. Issue will close unless more happens.
